### PR TITLE
Add support for upgrade_result command on task_manager

### PR DESCRIPTION
**Pull request by Nicogp**
|Related issue|
|---|
|#5511|

## Description

This PR adds the functionality to support upgrade_result command.
Translate to the upgrade status according the status of a task :

1. If the task's state is UPDATING, the upgrade module should return status UPDATING.
2. If the task's state is DONE, the upgrade module should return status UPDATED.
3. If the task's state is FAILED, the upgrade module should return status OUTDATED.
4. If the task's state is NEW or there isn't an entry for that agent, the upgrade module should return status ERROR.

Add fields create_time and update_time to identify when a task has started and has updated its status

## Logs/Alerts example
**Upgrade result request by agent_id:**
Request:
`[{"module": "api", "command": "upgrade_result", "agent": 10},{"module": "api", "command": "upgrade_result", "agent": 20}]`

Response:
`[{"error":0,"data":"Success","agent":10,"task_id":5,"module":"upgrade_module","command":"upgrade","status":"Outdated","create_time":"2020-08-06 23:45:56","update_time":0},{"error":8,"data":"No task for agent in DB","agent":20}]`

**Task result request by task_id**
Request:
`[{"module": "upgrade_module", "command": "task_result", "task_id": 1},{"module": "api", "command": "task_result", "task_id": 10}]`

Response:
`[{"error":0,"data":"Success","agent":1,"task_id":1,"module":"upgrade_module","command":"upgrade","status":"New","create_time":"2020-08-06 22:51:44","update_time":0},{"error":7,"data":"No task in DB","task_id":10}]`

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)

### CODE REVIEW

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 1]
```diff
@@ -252,6 +281,38 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
                 response = wm_task_manager_build_response(WM_TASK_INVALID_AGENT_ID, agent_id, task_id, status);
             }
 
+        } else if (!strcmp(commands_list[WM_TASK_UPGRADE_RESULT], command)) {
```
Comment (nicolaspapp) [inline src/wazuh_modules/task_manager/wm_task_manager.c:284]: Is there any way we can reduce code complexity on `wm_task_manager_analyze_task`? otherwise it will be really hard to do unit tests on it.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 2]
```diff
@@ -267,6 +328,56 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+        cJSON_AddStringToObject(res, json_keys[WM_TASK_MODULE], module);
+    }
+    if (command != NULL) {
+        cJSON_AddStringToObject(res, json_keys[WM_TASK_COMMAND], command);
+    }
+    if (status != NULL) {
+
+        if (!strcmp(commands_list[WM_TASK_UPGRADE_RESULT], request_command)){
+            
+            if (!strcmp(task_statuses[WM_TASK_DONE], status)){
```
Comment (nicolaspapp) [inline src/wazuh_modules/task_manager/wm_task_manager.c:343]: Is it any way we can map the task_status with the upgrade_status to avoid all this conditionals?

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 3]
```diff
@@ -267,6 +328,56 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+        char timestamp[40];
+        time_t tmp = create_time;
+        struct tm tm = { .tm_sec = 0 };
+        localtime_r(&tmp, &tm);
+        strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", &tm);
+        cJSON_AddStringToObject(res, json_keys[WM_TASK_CREATE_TIME], timestamp);
+    }
+    if (last_update_time != OS_INVALID) {
+        if (last_update_time > 0) {
+            char timestamp[40];
```
Comment (nicolaspapp) [inline src/wazuh_modules/task_manager/wm_task_manager.c:366]: timestamp generation is repeated, maybe can be done in a static method? I thnik that also the shared method `w_get_timestamp` could be used here

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 4]
```diff
@@ -59,7 +78,8 @@ static const char *error_codes[] = {
     [WM_TASK_INVALID_STATUS] = "Invalid status",
     [WM_TASK_DATABASE_NO_TASK] = "No task in DB",
     [WM_TASK_DATABASE_ERROR] = "Database error",
-    [WM_TASK_UNKNOWN_ERROR] = "Unknown error"
+    [WM_TASK_UNKNOWN_ERROR] = "Unknown error",
+    [WM_TASK_DATABASE_NO_AGENT] = "No task for agent in DB"
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:82]: This is not in the right order as in the enum.

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 5]
```diff
@@ -149,10 +152,6 @@ int wm_task_manager_get_task_status(int agent_id, const char *module, char **sta
     int result = 0;
     int task_id = OS_INVALID;
 
-    if (sqlite3_open_v2(TASKS_DB, &db, SQLITE_OPEN_READWRITE, NULL)) {
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:152]: Why did you remove this?

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 6]
```diff
@@ -280,4 +280,80 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+    switch (wdb_step(stmt)) {
+    case SQLITE_ROW:
+        task_id = sqlite3_column_int(stmt, 0);
+        sqlite_strdup((char*)sqlite3_column_text(stmt, 1), *command);
+        *create_time = sqlite3_column_int(stmt, 2);
+        *last_update_time = sqlite3_column_int(stmt, 3);
+        sqlite_strdup((char*)sqlite3_column_text(stmt, 4), *status);
+        result = task_id;
+        break;
+    case SQLITE_DONE:
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:308]: I think this case is not necessary.

Comment (TomasTurina) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c:308]: Check line 246, you should do this like that.

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 7]
```diff
@@ -280,4 +280,80 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+    if (sqlite3_open_v2(TASKS_DB, &db, SQLITE_OPEN_READWRITE, NULL)) {
+        mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_OPEN_DB_ERROR);
+        return wm_task_manager_sql_error(db, stmt);
+    }
+    if (wdb_prepare(db, task_queries[WM_TASK_GET_TASK], -1, &stmt, NULL) != SQLITE_OK) {
+        mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_SQL_PREPARE_ERROR);
+        return wm_task_manager_sql_error(db, stmt);
+    }
+    sqlite3_bind_int(stmt, 1, task_id);
+    switch (wdb_step(stmt)) {
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:335]: Idem here, check line 246.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 8]
```diff
@@ -202,9 +230,10 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
         os_strdup(tmp->valuestring, status);
     }
 
-    if (!strcmp(modules_list[WM_TASK_UPGRADE_MODULE], module)) {
+    if (!strcmp(modules_list[WM_TASK_UPGRADE_MODULE], module) || !strcmp(modules_list[WM_TASK_API_MODULE], module)) {
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:233]: I would change this for:

```c
if (!strcmp(modules_list[WM_TASK_UPGRADE_MODULE], module)) {

} else if (!strcmp(modules_list[WM_TASK_API_MODULE], module)) {

} else {

}
```

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 9]
```diff
@@ -252,6 +281,38 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
                 response = wm_task_manager_build_response(WM_TASK_INVALID_AGENT_ID, agent_id, task_id, status);
             }
 
+        } else if (!strcmp(commands_list[WM_TASK_UPGRADE_RESULT], command)) {
+            
+            if (agent_id != OS_INVALID) {
+                if (result = wm_task_manager_get_task_by_agent_id_and_module(agent_id, modules_list[WM_TASK_UPGRADE_MODULE], &command_result, &status, &create_time, &last_update_time), result == OS_INVALID){
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:287]: Use `task_id` instead of result.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 10]
```diff
@@ -252,6 +281,38 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+                    wm_task_manager_build_response_result(response, modules_list[WM_TASK_UPGRADE_MODULE], command_result, status, create_time, last_update_time, command);
+                    os_free(command_result);
+                }
+            } else {
+                *error_code = WM_TASK_INVALID_AGENT_ID;
+                response = wm_task_manager_build_response(WM_TASK_INVALID_AGENT_ID, agent_id, task_id, status);
+            }
+        } else if (!strcmp(commands_list[WM_TASK_TASK_RESULT], command)) {
+
+            if (result = wm_task_manager_get_task_by_task_id(task_id, &agent_id, &module_result, &command_result, &status, &create_time, &last_update_time), result == OS_INVALID){
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:304]: Idem here.

File: src/wazuh_modules/task_manager/wm_task_manager_db.c (hunk 5 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c:152]: Shouldn't have deleted it, error in the Merge

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 11]
```diff
@@ -110,6 +132,9 @@ size_t wm_task_manager_dispatch(const char *msg, char **response) {
         case WM_TASK_DATABASE_NO_TASK:
             mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_COULD_NOT_FIND_TASK, task);
             break;
+        case WM_TASK_DATABASE_NO_AGENT:
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:135]: This is repeated.

Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c:135]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 10 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: here I need to keep the two variables separate because if there was an error or it didn't find the task, it would lose the value of task_id to send the response message

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 12]
```diff
@@ -280,4 +284,85 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+
+    if (sqlite3_open_v2(TASKS_DB, &db, SQLITE_OPEN_READWRITE, NULL)) {
+        mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_OPEN_DB_ERROR);
+        return wm_task_manager_sql_error(db, stmt);
+    }
+    if (wdb_prepare(db, task_queries[WM_TASK_GET_TASK_BY_AGENT_ID_AND_MODULE], -1, &stmt, NULL) != SQLITE_OK) {
+        mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_SQL_PREPARE_ERROR);
+        return wm_task_manager_sql_error(db, stmt);
+    }
+    int task_id;
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:300]: Move this to the beginning of the function.

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 13]
```diff
@@ -280,4 +284,85 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+        *last_update_time = sqlite3_column_int(stmt, 5);
+        sqlite_strdup((char*)sqlite3_column_text(stmt, 6), *status);
+        result = task_id;
+        break;
+    case SQLITE_DONE:
+        result = OS_NOTFOUND;
+        break;
+    default:
+        mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_SQL_STEP_ERROR);
+        result = OS_INVALID;
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:359]: Replace this line with `return wm_task_manager_sql_error(db, stmt);`

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 14]
```diff
@@ -280,4 +284,85 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+    }
+
+    task_id = sqlite3_column_int(stmt, 0);
+
+    if (!task_id) {
+        wdb_finalize(stmt);
+        sqlite3_close_v2(db);
+        return OS_NOTFOUND;
+    }
+    sqlite_strdup((char*)sqlite3_column_text(stmt, 1), *command);
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:316]: Add this in an else so you can remove lines 312 and 313.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 15]
```diff
@@ -35,18 +36,37 @@ static const char *json_keys[] = {
     [WM_TASK_TASK_ID] = "task_id",
     [WM_TASK_STATUS] = "status",
     [WM_TASK_ERROR] = "error",
-    [WM_TASK_ERROR_DATA] = "data"
+    [WM_TASK_ERROR_DATA] = "data",
+    [WM_TASK_CREATE_TIME] = "create_time",
+    [WM_TASK_LAST_UPDATE_TIME] = "update_time"
+};
+
+static const char *task_statuses[] = {
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:44]: This is repeated in the file `wm_task_manager_db.c`. Remove it and add this line in the file `wm_task_manager.h`:
```c
extern const char* task_statuses[];
```

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 16]
```diff
@@ -60,6 +80,7 @@ static const char *error_codes[] = {
     [WM_TASK_DATABASE_NO_TASK] = "No task in DB",
     [WM_TASK_DATABASE_ERROR] = "Database error",
     [WM_TASK_UNKNOWN_ERROR] = "Unknown error"
+
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:83]: Extra line

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 17]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+        
+    } else if (!strcmp(modules_list[WM_TASK_API_MODULE], module)) {
+        response = wm_task_manager_analyze_task_api_module(task_object, error_code);
+        
+    } else {
+        *error_code = WM_TASK_INVALID_MODULE;
+        response = wm_task_manager_build_response(WM_TASK_INVALID_MODULE, agent_id, task_id, status);
+    }
+
+    if (status){
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:235]: `os_free` already does this check.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 18]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+    cJSON *response = NULL;
+    int result = 0;
+    int agent_id = OS_INVALID;
+    int task_id = OS_INVALID;
+    char *status = NULL;
+    
+    char *command = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_COMMAND])->valuestring;
+    char *module = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_MODULE])->valuestring;
+
+    if (tmp = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_AGENT_ID]), tmp && tmp->type == cJSON_Number) {
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:252]: This code is repeated. You should add `agent_id`, `task_id` and `status` as parameters of this function.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 19]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+    int task_id = OS_INVALID;
+    char *status = NULL;
+    int create_time = OS_INVALID;
+    int last_update_time = OS_INVALID;
+    char *command_result = NULL;
+    char *module_result = NULL;
+
+    char *command = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_COMMAND])->valuestring;
+
+    if (tmp = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_AGENT_ID]), tmp && tmp->type == cJSON_Number) {
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:335]: Idem here

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 20]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+cJSON* wm_task_manager_analyze_task_upgrade_module(const cJSON *task_object, int *error_code) {
+    cJSON *tmp = NULL;
+    cJSON *response = NULL;
+    int result = 0;
+    int agent_id = OS_INVALID;
+    int task_id = OS_INVALID;
+    char *status = NULL;
+    
+    char *command = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_COMMAND])->valuestring;
+    char *module = cJSON_GetObjectItem(task_object, json_keys[WM_TASK_MODULE])->valuestring;
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:250]: This is not necessary since you already know that this is the upgrade module.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 21]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
 
     } else {
-        *error_code = WM_TASK_INVALID_MODULE;
-        response = wm_task_manager_build_response(WM_TASK_INVALID_MODULE, agent_id, task_id, status);
+        *error_code = WM_TASK_INVALID_COMMAND;
+        response = wm_task_manager_build_response(WM_TASK_INVALID_COMMAND, agent_id, task_id, status);
     }
 
-    os_free(status);
+    if (status){
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:315]: Not necessary.

File: src/wazuh_modules/task_manager/wm_task_manager.h [hunk 22]
```diff
@@ -126,5 +159,18 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code);
...
+ * @param res JSON object response
+ * @param module Module of the task when receiving a request for a specific task.
+ * @param command Command of the task when receiving a request for a specific task.
+ * @param status Status of the task when receiving a request for a specific task.
+ * @param create_time Date of creation task.
+ * @param last_update_time Date of update task.
+ * @param request_command Command that requested the query.
+ * @return JSON object.
+ * */
+cJSON* wm_task_manager_build_response_result(cJSON *res, const char *module, const char *command, char *status, int create_time, int last_update_time, char *request_command);
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.h:173]: Change `res` for `response`.

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 23]
```diff
@@ -280,4 +284,85 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+    }
+    if (wdb_prepare(db, task_queries[WM_TASK_GET_TASK], -1, &stmt, NULL) != SQLITE_OK) {
+        mterror(WM_TASK_MANAGER_LOGTAG, MOD_TASK_SQL_PREPARE_ERROR);
+        return wm_task_manager_sql_error(db, stmt);
+    }
+    sqlite3_bind_int(stmt, 1, task_id);
+    
+    switch (wdb_step(stmt)) {
+    case SQLITE_ROW:
+        task_id = sqlite3_column_int(stmt, 0);
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:345]: This is not necessary

File: src/wazuh_modules/task_manager/wm_task_manager_db.c [hunk 24]
```diff
@@ -280,4 +284,85 @@ int wm_task_manager_update_task_status(int agent_id, const char *module, const c
...
+        *agent_id = sqlite3_column_int(stmt, 1);
+        sqlite_strdup((char*)sqlite3_column_text(stmt, 2), *module);
+        sqlite_strdup((char*)sqlite3_column_text(stmt, 3), *command);
+        *create_time = sqlite3_column_int(stmt, 4);
+        *last_update_time = sqlite3_column_int(stmt, 5);
+        sqlite_strdup((char*)sqlite3_column_text(stmt, 6), *status);
+        result = task_id;
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager_db.c:352]: I think this function should return the `agent_id` since it is searching using the `task_id`. Change it and update the function comment in the `.h` file. Also, remove `agent_id` from the parameters list.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 25]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+                wm_task_manager_build_response_result(response, modules_list[WM_TASK_UPGRADE_MODULE], command_result, status, create_time, last_update_time, command);
+                os_free(command_result);
+            }
+        } else {
+            *error_code = WM_TASK_INVALID_AGENT_ID;
+            response = wm_task_manager_build_response(WM_TASK_INVALID_AGENT_ID, agent_id, task_id, status);
+        }
+    } else if (!strcmp(commands_list[WM_TASK_TASK_RESULT], command)) {
+
+        if (result = wm_task_manager_get_task_by_task_id(task_id, &agent_id, &module_result, &command_result, &status, &create_time, &last_update_time), result == OS_INVALID){
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:365]: Use `agent_id` instead of `result`.

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 26]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+            wm_task_manager_build_response_result(response, modules_list[WM_TASK_UPGRADE_MODULE], command_result, status, create_time, last_update_time, command);
+            os_free(command_result);
+            os_free(module_result);
+        }
+    } else {
+        *error_code = WM_TASK_INVALID_COMMAND;
+        response = wm_task_manager_build_response(WM_TASK_INVALID_COMMAND, agent_id, task_id, status);
+    }
+
+    if (status){
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:382]: Not necessary

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 27]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+                response = wm_task_manager_build_response(WM_TASK_SUCCESS, agent_id, task_id, NULL);
+                wm_task_manager_build_response_result(response, modules_list[WM_TASK_UPGRADE_MODULE], command_result, status, create_time, last_update_time, command);
+                os_free(command_result);
+            }
+        } else {
+            *error_code = WM_TASK_INVALID_AGENT_ID;
+            response = wm_task_manager_build_response(WM_TASK_INVALID_AGENT_ID, agent_id, task_id, status);
+        }
+    } else if (!strcmp(commands_list[WM_TASK_TASK_RESULT], command)) {
+
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:364]: First check if `task_id` is not `OS_INVALID`:
```c
if (task_id != OS_INVALID) {
...
} else {
    *error_code = WM_TASK_INVALID_TASK_ID;
    response = wm_task_manager_build_response(WM_TASK_INVALID_TASK_ID, agent_id, task_id, status);
}
```

File: src/wazuh_modules/task_manager/wm_task_manager.c [hunk 28]
```diff
@@ -203,66 +222,203 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code) {
...
+        if (result = wm_task_manager_get_task_by_task_id(task_id, &agent_id, &module_result, &command_result, &status, &create_time, &last_update_time), result == OS_INVALID){
+            *error_code = WM_TASK_DATABASE_ERROR;
+            response = wm_task_manager_build_response(WM_TASK_INVALID_AGENT_ID, agent_id, result, status);
+        } else if (result == OS_NOTFOUND || result == 0) {
+            *error_code = WM_TASK_DATABASE_NO_TASK;
+            response = wm_task_manager_build_response(WM_TASK_DATABASE_NO_TASK, OS_INVALID, task_id, status);
+        } else {
+            response = wm_task_manager_build_response(WM_TASK_SUCCESS, agent_id, result, NULL);
+            wm_task_manager_build_response_result(response, modules_list[WM_TASK_UPGRADE_MODULE], command_result, status, create_time, last_update_time, command);
```
Comment (TomasTurina) [inline src/wazuh_modules/task_manager/wm_task_manager.c:373]: Since in this case the task could be anything, you should set the api as module and not the upgrade module.

File: src/wazuh_modules/task_manager/wm_task_manager.h [hunk 29]
```diff
@@ -50,6 +62,7 @@ typedef enum _error_code {
     WM_TASK_INVALID_TASK_ID,
     WM_TASK_INVALID_STATUS,
     WM_TASK_DATABASE_NO_TASK,
+    WM_TASK_DATABASE_NO_AGENT,
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.h]: removed

File: src/wazuh_modules/task_manager/wm_task_manager.h [hunk 30]
```diff
@@ -126,5 +159,18 @@ cJSON* wm_task_manager_analyze_task(const cJSON *task_object, int *error_code);
  * */
 cJSON* wm_task_manager_build_response(int error_code, int agent_id, int task_id, char *status);
 
+/**
+ * Add data to a JSON object response.
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.h:165]: done

File: src/wazuh_modules/task_manager/wm_task_manager_db.c (hunk 12 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c]: done

File: src/wazuh_modules/task_manager/wm_task_manager_db.c (hunk 13 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c]: done

File: src/wazuh_modules/task_manager/wm_task_manager_db.c (hunk 14 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c]: done

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 15 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: done

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 17 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 18 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 19 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 20 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 21 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.h (hunk 22 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.h]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager_db.c (hunk 23 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager_db.c (hunk 24 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager_db.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 25 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: done

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 26 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 27 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: done

File: src/wazuh_modules/task_manager/wm_task_manager.c (hunk 28 above)
Comment (Nicogp) [reply src/wazuh_modules/task_manager/wm_task_manager.c]: you're right
